src/cls/IPM/ResourceProcessor/WebApplication.cls

```diff
@@ -0,0 +1,263 @@
+/// then the default behavior for that resource will be bypassed in the current phase.
+Method OnPhase(pPhase As %String, ByRef pParams, Output pResourceHandled As %Boolean = 0) As %Status
+{
+    Set sc = $$$OK
+    Try {
+        If (pPhase = "Validate") {
+            If (..Name = "") {
+                Return $$$ERROR($$$GeneralError,"Name is required for web applications")
+            }
+            Set keyList = ""
+            Set key = ""
+            While 1 {
+                Set value = ..ResourceReference.Attributes.GetNext(.key)
+                If (key = "") {
+                    Quit
+                }
+                Set keyList = keyList _ $ListBuild(key)
+            }
+            Set badKeyList = ..DoPropertiesExist(keyList)
+            If ($ListLength(badKeyList) > 0) {
+                Set error = "The following invalid attributes were found for resource "_..ResourceReference.Name_": "_$ListToString(badKeyList)
+                $$$ThrowStatus($$$ERROR($$$GeneralError,error))
+            }
+        }
+    } Catch (ex) {
+        Set sc = ex.AsStatus()
+    }
+    Return sc
+}
+
+Method OnBeforePhase(pPhase As %String, ByRef pParams) As %Status
+{
+    /*
+     * NOTE: Web app creation is done in OnBeforePhase so it happens before other
+     * resource processors such as FileCopy run.
+     */
+    Set sc = $$$OK
+    Try {
+        $$$ThrowOnError(##super(pPhase,.pParams))
+        
+        Set tVerbose = $Get(pParams("Verbose"))
+        #; Set tIsApplication = ..ResourceReference.Module.Lifecycle.%IsA(##class(%ZHSLIB.PackageManager.Developer.Lifecycle.Application).%ClassName(1))
+        Set tIsApplication = ..ResourceReference.Module.Lifecycle.%IsA(##class(%IPM.Lifecycle.DeployedModule).%ClassName(1))
+        
+        // TODO: Make mirror-safe?
+        // Would require moving mirror-safe APIs to the package manager - at which point, why not just have the package manager manage the whole federation?
+        If ((pPhase = "Configure") || ((pPhase = "Activate") && tIsApplication)) {
+            // Create web application
+            Do ..CreateOrUpdateWebApp(tVerbose)
+        } ElseIf ((pPhase = "Unconfigure") || ((pPhase = "Clean") && tIsApplication)) {
+            // Remove web application
+            Do ..DeleteWebApp(tVerbose)
+        }
+    } Catch e {
+        Set sc = e.AsStatus()
+    }
+    Quit sc
+}
+
+Method CreateOrUpdateWebApp(pVerbose As %Boolean = 0) [ Internal ]
+{
+    // TODO: Make mirror-safe.
+    Set tOriginalNS = $namespace
+    Set dbDir = $$$defdir
+    New $namespace
+    Set $namespace = "%SYS"
+    
+    Merge properties = ..SecurityProperties
+    If $D(properties("MatchRoles"), tMatchRoles) # 2 {
+        Set properties("MatchRoles") = ..ReplaceMatchRoles(tMatchRoles, dbDir)
+    }
+    If $D(properties("NameSpace"), tNameSpace) {
+        Set properties("NameSpace") = ..ReplaceNameSpace(tNameSpace, tOriginalNS)
```

Comment:
No need for this since %Evaluate() handles it.

Reply:
Done!